src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala

```diff
@@ -341,6 +325,84 @@ class SamRecordClipper(val mode: ClippingMode, val autoClipAttributes: Boolean)
     }
   }
 
+  /** Clips extended attributes that are the same length as the bases were prior to clipping. */
+  protected def clipExtendedAttributes(rec: SamRecord, remove: Int, fromStart: Boolean): Unit = {
+    if (mode == ClippingMode.Hard && remove > 0 && autoClipAttributes) {
+      val newLength = rec.length
+      val oldLength = newLength + remove
+      rec.attributes.foreach {
+        case (tag, s: String)   if s.length == oldLength => rec(tag) = if (fromStart) s.drop(remove) else s.take(newLength)
+        case (tag, a: Array[_]) if a.length == oldLength => rec(tag) = if (fromStart) a.drop(remove) else a.take(newLength)
+        case _  => Unit
+      }
+    }
+  }
+
+  /**
+    * Ensures sufficient masking or hard clipping exists on reads that may have soft-clipping. The read will
+    * only be altered if:
+    *   1. ClippingMode is Hard of SoftWithMask
+    *   2. The read is already clipped at the appropriate end
+    *   3. Soft-clipping exist within the first `length` bases of the appropriate end
+    *
+    * If all of those conditions are met, the soft-clipping will be upgraded to masking or hard clipping
+    * up until length bases are clipped or masked. E.g. if the incoming cigar is `10H10S80M` and `length` is
+    * 15, the resulting cigar will be `15H5S80M`.
+    */
+  protected def upgradeClipping(rec: SamRecord, length: Int, fromStart: Boolean): Unit = if (mode != ClippingMode.Soft) {
+    require(length > 0, "Length must be >= 0")
+    def iter = if (fromStart) rec.cigar.iterator else rec.cigar.reverseIterator
+    val hardClipped = iter.takeWhile(_.operator == Op.H).map(_.length).sum
+    val softClipped = iter.dropWhile(_.operator == Op.H).takeWhile(_.operator == Op.S).map(_.length).sum
+
+    if (hardClipped < length && softClipped > 0) {
```

Comment:
How about just a quick comment for our future selves: `If not all the requested clipping is currently hard-clipped, and some soft-clipping exists.`

Reply:
Done!